packages/dcos-integration-test/extra/test_mesos.py

```diff
@@ -0,0 +1,140 @@
+import json
+import logging
+import uuid
+
+import requests
+
+from test_util.marathon import get_test_app
+from test_util.recordio import Decoder, Encoder
+
+
+def test_if_marathon_app_can_be_debugged(cluster):
+    def post(url, headers, json=None, data=None, stream=False):
+        r = requests.post(url, headers=headers, json=json, data=data, stream=stream)
+        logging.info(
+            'Got %s with POST request to %s with headers %s and json data %s.',
+            r.status_code,
+            url,
+            headers,
+            json
+        )
+        assert r.status_code == 200
+        return r
+
+    def find_container_id(state, app_id):
+        if 'frameworks' in state:
+            for framework in state['frameworks']:
+                if 'tasks' in framework:
+                    for task in framework['tasks']:
+                        if 'id' in task and app_id in task['id']:
+                            if 'statuses' in task and 'container_status' in task['statuses'][0]:
+                                if 'container_id' in task['statuses'][0]['container_status']:
+                                    container_status = task['statuses'][0]['container_status']
+                                    if 'container_id' in container_status:
+                                        if 'value' in container_status['container_id']:
+                                            return container_status['container_id']['value']
+
+    def find_agent_id(state, app_id):
+        if 'frameworks' in state:
+            for framework in state['frameworks']:
+                if 'tasks' in framework:
+                    for task in framework['tasks']:
+                        if 'id' in task and app_id in task['id']:
+                            if 'slave_id' in task:
+                                return task['slave_id']
+
+    def find_agent_hostname(state, agent_id):
+        if 'slaves' in state:
+            for agent in state['slaves']:
+                if 'id' in agent and agent['id'] == agent_id and 'hostname' in agent:
+                    return agent['hostname']
+
+    # Creates and yields the initial ATTACH_CONTAINER_INPUT message, then a data message,
+    # then an empty data chunk to indicate end-of-stream.
+    def _input_streamer(encoder, nested_container_id, input_data):
+        message = {'type': 'ATTACH_CONTAINER_INPUT', 'attach_container_input': {'type': 'CONTAINER_ID'}}
+        message['attach_container_input']['container_id'] = nested_container_id
+        yield encoder.encode(message)
```

Comment:
Hmm, didn't work on first attempt, but maybe you have some tips now that the rest of it's cleaned up.